File: Unknownbank/app/src/main/java/com/kasatria/kasatriaunknownbank/CreditCardPersonalDetailsScreen.kt
package com.kasatria.kasatriaunknownbank

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

import com.kasatria.kasatriaunknownbank.ui.theme.PrimaryBlue
import com.kasatria.kasatriaunknownbank.ui.theme.TextPrimary
import com.kasatria.kasatriaunknownbank.ui.theme.White


@Composable
fun CreditCardPersonalDetailsScreen(
    productName: String,
    applicationData: CreditCardApplicationData,
    onApplicationDataChange: (CreditCardApplicationData) -> Unit,
    onBack: () -> Unit,
    onNext: () -> Unit
) {

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(White)
    ) {

        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
        ) {

            /*
             * =================================
             * HEADER
             * =================================
             */

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(56.dp)
            ) {

                IconButton(
                    onClick = onBack,

                    modifier = Modifier
                        .padding(start = 8.dp)
                        .align(Alignment.CenterStart)
                        .size(48.dp)
                ) {

                    Icon(
                        imageVector =
                            Icons.AutoMirrored.Filled.ArrowBack,

                        contentDescription = "Back",

                        tint = TextPrimary
                    )
                }


                Text(
                    text =
                        "Applying for $productName",

                    modifier = Modifier
                        .align(Alignment.Center)
                        .padding(
                            start = 55.dp,
                            end = 30.dp
                        ),

                    color = Color(0xFF333333),

                    fontSize = 13.sp,

                    fontWeight =
                        FontWeight.Bold,

                    maxLines = 1
                )
            }


            /*
             * =================================
             * PERSONAL DETAILS + PROGRESS
             * =================================
             */

            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(
                        start = 20.dp,
                        end = 23.dp,
                        top = 20.dp
                    ),

                horizontalArrangement =
                    Arrangement.SpaceBetween,

                verticalAlignment =
                    Alignment.CenterVertically
            ) {

                Text(
                    text = "Personal Details",

                    color = Color(0xFF333333),

                    fontSize = 15.sp,

                    fontWeight =
                        FontWeight.Bold
                )


                Text(
                    text = "20%",

                    color =
                        Color(0xFF0EAF5F),

                    fontSize = 14.sp,

                    fontWeight =
                        FontWeight.Medium
                )
            }


            Spacer(
                modifier =
                    Modifier.height(11.dp)
            )


            /*
             * Progress track
             */

            Box(
                modifier = Modifier
                    .padding(
                        horizontal = 20.dp
                    )
                    .fillMaxWidth()
                    .height(6.dp)
                    .background(
                        color =
                            Color(0xFFE6E8EA),

                        shape =
                            RoundedCornerShape(100.dp)
                    )
            ) {

                Box(
                    modifier = Modifier
                        .fillMaxWidth(0.20f)
                        .height(6.dp)
                        .background(
                            color =
                                Color(0xFF0EAF5F),

                            shape =
                                RoundedCornerShape(100.dp)
                        )
                )
            }


            Spacer(
                modifier =
                    Modifier.height(23.dp)
            )


            /*
             * =================================
             * FORM
             * =================================
             */

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(
                        horizontal = 20.dp
                    ),

                verticalArrangement =
                    Arrangement.spacedBy(18.dp)
            ) {

                SelectionField(
                    label = "Title",
                    value = applicationData.title,
                    options = listOf(
                        "MS / PUAN",
                        "MR / ENCIK"
                    ),
                    onSelected = {
                       onApplicationDataChange(
                           applicationData.copy(
                               title = it
                           )
                       )
                    }
                )


                PersonalField(
                    label =
                        "Full Name (Completes as per NRIC)",

                    value = applicationData.fullName,

                    onValueChange = {
                        onApplicationDataChange(
                            applicationData.copy(
                                fullName = it
                            )
                        )
                    }
                )


                PersonalField(
                    label = "IC No.",

                    value = applicationData.icNumber,

                    onValueChange = {
                        onApplicationDataChange(
                            applicationData.copy(
                                icNumber = it
                            )
                        )
                    }
                )


                PersonalField(
                    label = "Date of birth",

                    value = applicationData.dateOfBirth,

                    onValueChange = {
                        onApplicationDataChange(
                            applicationData.copy(
                                dateOfBirth = it
                            )
                        )
                    }
                )


                PersonalField(
                    label = "Email",

                    value = applicationData.email,

                    onValueChange = {
                        onApplicationDataChange(
                            applicationData.copy(
                                email = it
                            )
                        )
                    }
                )


                /*
                 * Mobile No.
                 */

                Column {

                    Text(
                        text = "Mobile No.",

                        color =
                            Color(0xFF333333),

                        fontSize = 14.sp
                    )


                    Spacer(
                        modifier =
                            Modifier.height(7.dp)
                    )


                    Row(
                        horizontalArrangement =
                            Arrangement.spacedBy(10.dp)
                    ) {

                        SelectionBox(
                            value = applicationData.phoneCountryCode,

                            options = listOf(
                                "+60",
                                "+65",
                                "+62"
                            ),

                            onSelected = {
                                onApplicationDataChange(
                                    applicationData.copy(
                                        phoneCountryCode = it
                                    )
                                )
                            },

                            modifier =
                                Modifier.width(100.dp)
                        )


                        BasicFormField(
                            value = applicationData.phoneNumber,

                            onValueChange = {
                                onApplicationDataChange(
                                    applicationData.copy(
                                        phoneNumber = it
                                    )
                                )
                            },

                            modifier =
                                Modifier.weight(1f)
                        )
                    }
                }
            }
        }


        /*
         * =================================
         * NEXT
         * =================================
         */

        Button(
            onClick = onNext,

            modifier = Modifier
                .align(Alignment.BottomCenter)
                .navigationBarsPadding()
                .padding(
                    start = 16.dp,
                    end = 16.dp,
                    bottom = 16.dp
                )
                .fillMaxWidth()
                .height(46.dp),

            shape =
                RoundedCornerShape(100.dp),

            colors =
                ButtonDefaults.buttonColors(
                    containerColor =
                        PrimaryBlue,

                    contentColor =
                        White
                )
        ) {

            Text(
                text = "Next",

                fontSize = 15.sp,

                fontWeight =
                    FontWeight.Bold
            )
        }
    }
}


@Composable
private fun PersonalField(
    label: String,
    value: String,
    onValueChange: (String) -> Unit
) {

    Column {

        Text(
            text = label,

            color =
                Color(0xFF333333),

            fontSize = 14.sp
        )


        Spacer(
            modifier =
                Modifier.height(7.dp)
        )


        BasicFormField(
            value = value,

            onValueChange =
                onValueChange,

            modifier =
                Modifier.fillMaxWidth()
        )
    }
}


@Composable
fun BasicFormField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    backgroundColor: Color = Color(0xFFF7F9FB)
) {

    val shape =
        RoundedCornerShape(8.dp)

    Box(
        modifier = modifier
            .height(46.dp)
            .background(
                color =
                    Color(0xFFF7F9FB),

                shape = shape
            )
            .border(
                width = 1.dp,

                color =
                    Color(0xFFDDDDDD),

                shape = shape
            )
            .padding(
                horizontal = 19.dp
            ),

        contentAlignment =
            Alignment.CenterStart
    ) {

        BasicTextField(
            value = value,

            onValueChange =
                onValueChange,

            singleLine = true,

            modifier =
                Modifier.fillMaxWidth(),

            textStyle =
                TextStyle(
                    color =
                        Color(0xFF666666),

                    fontSize = 15.sp
                )
        )
    }
}


@Composable
private fun SelectionField(
    label: String,
    value: String,
    options: List<String>,
    onSelected: (String) -> Unit
) {

    Column {

        Text(
            text = label,

            color =
                Color(0xFF333333),

            fontSize = 14.sp
        )


        Spacer(
            modifier =
                Modifier.height(7.dp)
        )


        SelectionBox(
            value = value,
            options = options,
            onSelected = onSelected,
            modifier =
                Modifier.fillMaxWidth()
        )
    }
}


@Composable
fun SelectionBox(
    value: String,
    options: List<String>,
    onSelected: (String) -> Unit,
    modifier: Modifier = Modifier,
    backgroundColor: Color = Color(0xFFF7F9FB)
) {

    var expanded by rememberSaveable {
        mutableStateOf(false)
    }

    val shape =
        RoundedCornerShape(8.dp)


    Box(
        modifier = modifier
    ) {

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(46.dp)
                .background(
                    color =
                        Color(0xFFF7F9FB),

                    shape = shape
                )
                .border(
                    width = 1.dp,

                    color =
                        Color(0xFFDDDDDD),

                    shape = shape
                )
                .clickable {
                    expanded = true
                }
                .padding(
                    start = 19.dp,
                    end = 10.dp
                ),

            verticalAlignment =
                Alignment.CenterVertically,

            horizontalArrangement =
                Arrangement.SpaceBetween
        ) {

            Text(
                text = value,

                color =
                    Color(0xFF666666),

                fontSize = 15.sp
            )


            Icon(
                imageVector =
                    Icons.Default.KeyboardArrowDown,

                contentDescription =
                    "Select",

                tint =
                    Color(0xFF777777),

                modifier =
                    Modifier.size(24.dp)
            )
        }


        DropdownMenu(
            expanded = expanded,

            onDismissRequest = {
                expanded = false
            }
        ) {

            options.forEach { option ->

                DropdownMenuItem(
                    text = {
                        Text(option)
                    },

                    onClick = {
                        onSelected(option)
                        expanded = false
                    }
                )
            }
        }
    }
}

File: unknownbank_flutter/lib/screens/credit_card_personal_details_screen.dart
import 'package:flutter/material.dart';

import '../models/credit_card_application_data.dart';
import '../theme/app_colors.dart';

class CreditCardPersonalDetailsScreen
    extends StatefulWidget {
  final String productName;
  final CreditCardApplicationData applicationData;

  final ValueChanged<CreditCardApplicationData>
      onApplicationDataChange;

  final VoidCallback onNext;

  const CreditCardPersonalDetailsScreen({
    super.key,
    required this.productName,
    required this.applicationData,
    required this.onApplicationDataChange,
    required this.onNext,
  });

  @override
  State<CreditCardPersonalDetailsScreen>
      createState() =>
          _CreditCardPersonalDetailsScreenState();
}

class _CreditCardPersonalDetailsScreenState
    extends State<CreditCardPersonalDetailsScreen> {
  late CreditCardApplicationData data;

  @override
  void initState() {
    super.initState();
    data = widget.applicationData;
  }

  void _update(
    CreditCardApplicationData newData,
  ) {
    setState(() {
      data = newData;
    });

    widget.onApplicationDataChange(newData);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Column(
          children: [
            SizedBox(
              height: 56,
              child: Stack(
                alignment: Alignment.center,
                children: [
                  Align(
                    alignment: Alignment.centerLeft,
                    child: Padding(
                      padding:
                          const EdgeInsets.only(left: 8),
                      child: IconButton(
                        onPressed: () {
                          Navigator.of(context).pop();
                        },
                        icon: const Icon(
                          Icons.arrow_back,
                          color: AppColors.textPrimary,
                        ),
                      ),
                    ),
                  ),

                  Padding(
                    padding: const EdgeInsets.only(
                      left: 55,
                      right: 30,
                    ),
                    child: Text(
                      'Applying for ${widget.productName}',
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                      style: const TextStyle(
                        fontSize: 13,
                        fontWeight: FontWeight.bold,
                        color: Color(0xFF333333),
                      ),
                    ),
                  ),
                ],
              ),
            ),

            Padding(
              padding: const EdgeInsets.fromLTRB(
                20,
                20,
                23,
                0,
              ),
              child: const Row(
                mainAxisAlignment:
                    MainAxisAlignment.spaceBetween,
                children: [
                  Text(
                    'Personal Details',
                    style: TextStyle(
                      fontSize: 15,
                      fontWeight: FontWeight.bold,
                      color: Color(0xFF333333),
                    ),
                  ),

                  Text(
                    '20%',
                    style: TextStyle(
                      fontSize: 14,
                      fontWeight: FontWeight.w500,
                      color: Color(0xFF0EAF5F),
                    ),
                  ),
                ],
              ),
            ),

            const SizedBox(height: 11),

            Padding(
              padding: const EdgeInsets.symmetric(
                horizontal: 20,
              ),
              child: ClipRRect(
                borderRadius:
                    BorderRadius.circular(100),
                child: LinearProgressIndicator(
                  value: 0.20,
                  minHeight: 6,
                  backgroundColor:
                      const Color(0xFFE6E8EA),
                  valueColor:
                      const AlwaysStoppedAnimation(
                    Color(0xFF0EAF5F),
                  ),
                ),
              ),
            ),

            const SizedBox(height: 23),

            Expanded(
              child: SingleChildScrollView(
                padding: const EdgeInsets.fromLTRB(
                  20,
                  0,
                  20,
                  90,
                ),
                child: Column(
                  children: [
                    _DropdownField(
                      label: 'Title',
                      value: data.title,
                      options: const [
                        'MS / PUAN',
                        'MR / ENCIK',
                      ],
                      onChanged: (value) {
                        _update(
                          data.copyWith(
                            title: value,
                          ),
                        );
                      },
                    ),

                    const SizedBox(height: 18),

                    _TextField(
                      label:
                          'Full Name (Completes as per NRIC)',
                      value: data.fullName,
                      onChanged: (value) {
                        _update(
                          data.copyWith(
                            fullName: value,
                          ),
                        );
                      },
                    ),

                    const SizedBox(height: 18),

                    _TextField(
                      label: 'IC No.',
                      value: data.icNumber,
                      onChanged: (value) {
                        _update(
                          data.copyWith(
                            icNumber: value,
                          ),
                        );
                      },
                    ),

                    const SizedBox(height: 18),

                    _TextField(
                      label: 'Date of birth',
                      value: data.dateOfBirth,
                      onChanged: (value) {
                        _update(
                          data.copyWith(
                            dateOfBirth: value,
                          ),
                        );
                      },
                    ),

                    const SizedBox(height: 18),

                    _TextField(
                      label: 'Email',
                      value: data.email,
                      onChanged: (value) {
                        _update(
                          data.copyWith(
                            email: value,
                          ),
                        );
                      },
                    ),

                    const SizedBox(height: 18),

                    Column(
                      crossAxisAlignment:
                          CrossAxisAlignment.start,
                      children: [
                        const Text(
                          'Mobile No.',
                          style: TextStyle(
                            fontSize: 14,
                            color:
                                Color(0xFF333333),
                          ),
                        ),

                        const SizedBox(height: 7),

                        Row(
                          children: [
                            SizedBox(
                              width: 100,
                              child: _DropdownBox(
                                value: data
                                    .phoneCountryCode,
                                options: const [
                                  '+60',
                                  '+65',
                                  '+62',
                                ],
                                onChanged: (value) {
                                  _update(
                                    data.copyWith(
                                      phoneCountryCode:
                                          value,
                                    ),
                                  );
                                },
                              ),
                            ),

                            const SizedBox(width: 10),

                            Expanded(
                              child: _TextBox(
                                value:
                                    data.phoneNumber,
                                keyboardType:
                                    TextInputType
                                        .phone,
                                onChanged: (value) {
                                  _update(
                                    data.copyWith(
                                      phoneNumber:
                                          value,
                                    ),
                                  );
                                },
                              ),
                            ),
                          ],
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),

      bottomNavigationBar: SafeArea(
        top: false,
        child: Padding(
          padding: const EdgeInsets.fromLTRB(
            16,
            8,
            16,
            16,
          ),
          child: SizedBox(
            height: 46,
            child: FilledButton(
              onPressed: widget.onNext,
              style: FilledButton.styleFrom(
                backgroundColor:
                    AppColors.primary,
                shape: RoundedRectangleBorder(
                  borderRadius:
                      BorderRadius.circular(100),
                ),
              ),
              child: const Text(
                'Next',
                style: TextStyle(
                  fontSize: 15,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class _TextField extends StatelessWidget {
  final String label;
  final String value;
  final ValueChanged<String> onChanged;

  const _TextField({
    required this.label,
    required this.value,
    required this.onChanged,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment:
          CrossAxisAlignment.start,
      children: [
        Text(
          label,
          style: const TextStyle(
            fontSize: 14,
            color: Color(0xFF333333),
          ),
        ),

        const SizedBox(height: 7),

        _TextBox(
          value: value,
          onChanged: onChanged,
        ),
      ],
    );
  }
}

class _TextBox extends StatefulWidget {
  final String value;
  final ValueChanged<String> onChanged;
  final TextInputType? keyboardType;

  const _TextBox({
    required this.value,
    required this.onChanged,
    this.keyboardType,
  });

  @override
  State<_TextBox> createState() =>
      _TextBoxState();
}

class _TextBoxState extends State<_TextBox> {
  late final TextEditingController controller;

  @override
  void initState() {
    super.initState();

    controller =
        TextEditingController(text: widget.value);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: 46,
      child: TextField(
        controller: controller,
        keyboardType: widget.keyboardType,
        onChanged: widget.onChanged,
        style: const TextStyle(
          color: Color(0xFF666666),
          fontSize: 15,
        ),
        decoration: InputDecoration(
          filled: true,
          fillColor: const Color(0xFFF7F9FB),
          contentPadding:
              const EdgeInsets.symmetric(
            horizontal: 19,
          ),
          border: OutlineInputBorder(
            borderRadius:
                BorderRadius.circular(8),
            borderSide: const BorderSide(
              color: Color(0xFFDDDDDD),
            ),
          ),
          enabledBorder: OutlineInputBorder(
            borderRadius:
                BorderRadius.circular(8),
            borderSide: const BorderSide(
              color: Color(0xFFDDDDDD),
            ),
          ),
        ),
      ),
    );
  }
}

class _DropdownField extends StatelessWidget {
  final String label;
  final String value;
  final List<String> options;
  final ValueChanged<String> onChanged;

  const _DropdownField({
    required this.label,
    required this.value,
    required this.options,
    required this.onChanged,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment:
          CrossAxisAlignment.start,
      children: [
        Text(
          label,
          style: const TextStyle(
            fontSize: 14,
            color: Color(0xFF333333),
          ),
        ),

        const SizedBox(height: 7),

        _DropdownBox(
          value: value,
          options: options,
          onChanged: onChanged,
        ),
      ],
    );
  }
}

class _DropdownBox extends StatelessWidget {
  final String value;
  final List<String> options;
  final ValueChanged<String> onChanged;

  const _DropdownBox({
    required this.value,
    required this.options,
    required this.onChanged,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 46,
      padding: const EdgeInsets.only(
        left: 19,
        right: 10,
      ),
      decoration: BoxDecoration(
        color: const Color(0xFFF7F9FB),
        borderRadius: BorderRadius.circular(8),
        border: Border.all(
          color: const Color(0xFFDDDDDD),
        ),
      ),
      child: DropdownButtonHideUnderline(
        child: DropdownButton<String>(
          value: value,
          isExpanded: true,
          icon: const Icon(
            Icons.keyboard_arrow_down,
            color: Color(0xFF777777),
          ),
          style: const TextStyle(
            color: Color(0xFF666666),
            fontSize: 15,
          ),
          items: options
              .map(
                (option) =>
                    DropdownMenuItem<String>(
                  value: option,
                  child: Text(option),
                ),
              )
              .toList(),
          onChanged: (value) {
            if (value != null) {
              onChanged(value);
            }
          },
        ),
      ),
    );
  }
}